pkg/image/registryclient/alternativesources.go

```diff
@@ -0,0 +1,193 @@
+package registryclient
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+
+	"github.com/docker/distribution"
+	operatorv1alpha1 "github.com/openshift/api/operator/v1alpha1"
+	operatorv1alpha1scheme "github.com/openshift/client-go/operator/clientset/versioned/scheme"
+	operatorv1alpha1client "github.com/openshift/client-go/operator/clientset/versioned/typed/operator/v1alpha1"
+	imagereference "github.com/openshift/library-go/pkg/image/reference"
+)
+
+// AlternativeImageSources holds ImageContentSourcePolicy variables to look up image sources
+type AlternativeImageSources struct {
+	ICSPFile     string
+	ICSPList     []operatorv1alpha1.ImageContentSourcePolicy
+	ICSPClientFn func() (operatorv1alpha1client.ImageContentSourcePolicyInterface, error)
+}
+
+// WithICSP returns a Context with ImageContentSourcePolicy variables populated
+func (c *Context) WithICSP(icspFile string, icspClientFn func() (operatorv1alpha1client.ImageContentSourcePolicyInterface, error)) *Context {
+	c.ImageSources = AlternativeImageSources{
+		ICSPFile:     icspFile,
+		ICSPClientFn: icspClientFn,
+	}
+	return c
+}
+
+// addICSPsFromCluster will lookup ICSPs from cluster. Logs errors rather than returning errors. By design,
+// this function has no way of knowing whether it expects to connect to a cluster or not. That is up the the caller.
+func (a *Context) addICSPsFromCluster(icspClient operatorv1alpha1client.ImageContentSourcePolicyInterface) {
+	if icspClient == nil {
+		klog.V(4).Infof("no client exists to access ImageContentSourcePolicies in cluster")
+		return
+	}
+	icsps, err := icspClient.List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		// may or may not have access to ICSPs in cluster
+		// don't error if can't access ICSPs
+		klog.V(4).Infof("did not access any ImageContentSourcePolicies in cluster: %v", err)
+	}
+	if len(icsps.Items) == 0 {
+		// might be looking up implicitly, log only
+		klog.V(4).Info("no ImageContentSourcePolicies found in cluster")
+	}
+	a.ImageSources.ICSPList = append(a.ImageSources.ICSPList, icsps.Items...)
+}
+
+// addImageSourcePoliciesFromFile appends to list of alternative image sources from ICSP file
+func (a *Context) addImageSourcePoliciesFromFile() error {
```

Comment:
If you have separate function for reading, I'd move here also the check for `ICSPFile` nil.

Reply:
right, updated